### DISCUSSION TITLE
Cannot compile gleam_stdlib when using mix_gleam

### DISCUSSION THREAD

**Discussion by mradke** [gleam-lang/gleam#3483] [Questions & support]
I’m currently trying to test gleam inside of a mix project. I followed the [GitHub - gleam-lang/mix_gleam: ⚗️ Build Gleam code with mix](https://github.com/gleam-lang/mix_gleam) readme but when I try to compile I’ll get the following error:

`Could not compile :gleam_stdlib, no "mix.exs", "rebar.config" or "Makefile" (pass :compile as an option to customize compilation, set it to "false" to do nothing)`

The configuration in mix.exs looks like this:

```exs
...
  @app :gleam_exp

  def project do
    [
      app: @app,
      version: "0.1.0",
      elixir: "~> 1.14",
      elixirc_paths: elixirc_paths(Mix.env()),
      start_permanent: Mix.env() == :prod,
      aliases: aliases(),

      # Gleam config
      archives: [mix_gleam: "~> 0.6"],
      compilers: [:gleam | Mix.compilers()],
      aliases: [
        # Or add this to your aliases function
        "deps.get": ["deps.get", "gleam.deps.get"]
      ],
      erlc_paths: [
        "build/dev/erlang/#{@app}/_gleam_artefacts",
        # For Gleam < v0.25.0
        "build/dev/erlang/#{@app}/build"
      ],
      erlc_include_path: "build/dev/erlang/#{@app}/include",
      # For Elixir >= v1.15.0
      prune_code_paths: false,

      deps: deps(),
    ]
  end
...
```

Elixir version: 1.16.3 (compiled with Erlang/OTP 25)
Gleam version: 1.3.2

Any pointers why this is happening?

**Comment by dvic** [comment]
And you added gleam_stdlib to the mix deps?

```elixir
# in mix.exs
# ...
  defp deps do
    [
      # ...
      {:gleam_stdlib, "~> 0.34 or ~> 1.0"},
      {:gleeunit, "~> 1.0", only: [:dev, :test], runtime: false},
      # ...
    ]
  end
# ...
```

**Reply by mradke** [reply]
Thanks for the answer! Yes. I followed the guide and added the deps. They are also in the deps folder locally but they are indeed missing the `"mix.exs", "rebar.config" or "Makefile"` files.

**Reply by dvic** [reply]
And what about the example project? https://github.com/gleam-lang/mix_gleam/tree/main/test_projects/basic_project

I just checked that one and `mix test` runs fine on my machine (OTP 27 + Elxir 1.17) (after running `mix deps.get`).

**Comment by zakw** [comment]
Hello, I believe I have a reproduction of the issue that @mradke ran into.

For me, it was in the context of attempting to add Gleam to an existing Phoenix application. The default Phoenix template has a slightly more involved `mix.exs` than a default `mix new` project so perhaps that's what was masking the discovery of this issue.

Here is a simplified repo with each commit representing the set of steps to reproduce and subsequently "fix" the issue with a hack:
https://github.com/zakw/gleam_mix_test/commits/main/

# Steps to Reproduce
1. Ensure Phoenix is installed `mix archive.install hex phx_new` (see https://hexdocs.pm/phoenix/installation.html)
2. Create a new default Phoenix project `mix phx.new gleam_test --database sqlite3`
3. Add Gleam compiler and dependencies support following the standard [README.md](https://github.com/gleam-lang/mix_gleam/blob/main/README.md)
4. Run `mix setup`

**Note:** In-between test attempts I would run `git clean -xdf` to ensure I had a clean, first-time environment before running `mix setup` again.

# Observations
- The first time `mix setup` is invoked, the `gleam_stdlib` fails to compile. It appears to compile after a subsequent invocation of `mix setup` and continue with the application first-time setup.
- If you instead invoke `mix deps.get` and then run `mix setup`, separately, it builds and runs without issue even though `setup` invokes `deps.get`, itself
- No amount of tweaking or adding `compile`, `deps.compile`, `deps.compile gleam_stdlib` to the various aliases seemed to avoid the first-time compilation issue

# The "Fix"
My hack fix was to modify the `gleam.deps.get` command to `cmd mix gleam.deps.get` to run it in a separate process. My sneaking suspicion is that either:
- a) there is some post-invocation magic that `gleam.deps.get` does which is not properly captured by the `setup` alias pipeline or
- b) there is some environment issue induced by `gleam.deps.get` that is avoided by running it in a separate process

This "hack fix" is illustrated, here:
https://github.com/zakw/gleam_mix_test/commit/1be6c9c91e0b658bca5a3bc9e70396c876418e42

I'm unsure what the true underlying issue is or what the appropriate solution is, but I thought I'd illustrate my findings here in case that helps point someone who knows more about this system in a good direction.

# System Environment
- Windows 10 22H2 x64
- Powershell 7.5.4
- VS Code 1.107.1
- Erlang/OTP 28
- Elixir 1.19.4
- Phoenix 1.8.3
- Gleam 1.14.0

(_Copied to https://github.com/gleam-lang/mix_gleam/issues/44_)
